httpcore/_async/http_proxy.py

```diff
@@ -140,47 +133,52 @@ def __init__(
         connection = await self._get_connection_from_pool(origin)
 
         if connection is None:
-            connection = AsyncHTTPConnection(
-                origin=origin, http2=False, ssl_context=self._ssl_context,
+            # First, create a connection to the proxy server
+            proxy_connection = AsyncHTTPConnection(
+                origin=self.proxy_origin, http2=False, ssl_context=self._ssl_context,
             )
             async with self._thread_lock:
                 self._connections.setdefault(origin, set())
-                self._connections[origin].add(connection)
+                self._connections[origin].add(proxy_connection)
 
-            # Establish the connection by issuing a CONNECT request...
+            # Issue a CONNECT request...
 
             # CONNECT www.example.org:80 HTTP/1.1
             # [proxy-headers]
             target = b"%b:%d" % (url[1], url[2])
             connect_url = self.proxy_origin + (target,)
-            connect_headers = self.proxy_headers
-            proxy_response = await connection.request(
-                b"CONNECT", connect_url, headers=connect_headers, timeout=timeout
+            headers = self.proxy_headers + ([] if headers is None else headers)
```

Comment:
Yeah, you're right. Looks like in HTTPX we were [copying over the proxy headers and adding an additional `Accept` header](https://github.com/encode/httpx/blob/0.12.1/httpx/_dispatch/proxy_http.py#L140-L141), I'll just go with that.

Reply:
I'm actually not sure we need `Accept` either. Neither [RFC 7321](https://tools.ietf.org/html/rfc7231#section-4.3.6) nor the [MDN](https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/CONNECT) mention it in examples or specs. WDYT?

Reply:
From what I can see we had not discussed this point back in https://github.com/encode/httpx/pull/259 either, though I may have missed a comment thread.

Should we leave it out and see if it breaks things?

Reply:
This turned out to be slightly more complex that I would've liked.

In forwarding we do copy the request headers, which makes passing `host` in the proxy headers unnecessary. Forcing the user to pass the `host` header to `AsyncHTTPProxy` only when tunnelling is not great, particularly when default mode choosed which method to use based on the URL. So I wrote a function to pull the `host` from the request header if not present and add the `accept: */*` to keep things consistent.

Reply:
I don't think there should be any smarts built into HTTPCore — this was one of the major points of extracting this networking layer to a separate library, and leaving smarts to the client in HTTPX or other users of HTTPCore.

It makes sense to me that users should pass the `Host` when they're planning to use tunnel proxying, because it's required by the spec to perform the tunnel setup.

Similarly, it makes sense to me that users don't have to pass a proxy `Host` when doing forwarding, because as you noted the host is already provided at the request level, and the proxy doesn't need to provide its `Host` at all.

I really think we're all good as far as HTTPCore is concerned, and the for-convenience "let's include headers from the request because it's cumbersome to have to add proxy host headers" should be up to HTTPX.

Does that make sense?

I'm also still not convinced we need `accept: */*` to be added here (not sure what function it serves in making the proxy requests), but it'd probably be okay to add it in HTTPX if what we want is consistency of behavior at the client level.

HTTPCore should be as tightly-scoped as possible: just do what's acceptable by the HTTP spec, and set constraints on inputs accordingly. It's one of its core promises IMO.

Reply:
It does make sense, but in my mind by having a `DEFAULT` proxy mode we are unifying the APIs for tunneling and forwarding for that case, hiding the fact that we're using one or the other from the user.

If we did not have the default mode I'd agree with you. Maybe the default mode is not in the spirit of HTTPCore?

Reply:
I'm leaning toward removing the `DEFAULT` mode in HTTPCore and moving the logic for checking which mode should be used and putting together the required headers there.

> I'm also still not convinced we need accept: */* to be added here

@sethmlarson [included it](https://github.com/encode/httpx/commit/57c27cac36f9ca28201e26e35ae4f79be2051a8e#diff-a4cfa44ae0ce9e3e95aae35ee88f1cb9R119) back when he first added proxy support but it was lost in the move to HTTPCore. I honestly can't say why  but I'm sure Seth had a good reason 🙂

Reply:
https://github.com/encode/httpcore/pull/57/commits/7066bf0adf1ebfd44e03da55c2154b83378d6d97

Let me know what you think, resolving this discussion 👍

Reply:
FYI https://github.com/urllib3/urllib3/blob/29b214a129883301f91ae4a74fd7ef2958cbf7b0/src/urllib3/poolmanager.py#L484